theano/gof/wrapper.py

```diff
@@ -0,0 +1,419 @@
+            self.attr2 = value_attr2
+
+    In perform() implementation (with params named `param`):
+
+        var1 = param.attr1
+        var2 = param.attr2
+
+    In c_code() implementation (with `param = sub['params']`):
+
+        PyArrayObject* attr1 = param.attr1;
+        PyArrayObject* attr2 = param.attr2;
+        /* You won't need to free them or whatever else. */
+
+
+See `theano/gof/tests/test_wrapper.py` for a complete working example.
+
+"""
+
+from __future__ import absolute_import, print_function, division
+import re
+import hashlib
+import numpy
+from theano.gof.utils import MethodNotDefined
+from theano.gof import Type
+from theano.tensor.utils import hash_from_ndarray
+
+# NB: Maybe we should check if an attribute name is a C/C++ keyword, and raise an error if so.
+# These are some lists of C/C++ keywords:
+# http://fr.cppreference.com/w/cpp/keyword
+# http://fr.cppreference.com/w/c/keyword
+
+
+class Wrap(dict):
+    """
+    Internal convenient class to wrap many Python objects into one
+    (this class is not safe as the hash method does not check if values are effectively hashable).
+
+    Example:
+        >>> w = Wrap(attr1=1, attr2=2.0, attri='3')
+        >>> print(w.attr1, w.attr2, w.attri)
+        >>> d = dict(a=1, b=2, c='test')
+        >>> w2 = Wrap(**d)
+        >>> print(w2.a, w2.b, w2.c)
+
+    """
+
+    def __init__(self, **kwargs):
+        super(Wrap, self).__init__(**kwargs)
+        if len(kwargs) == 0:
+            raise TypeError('Wrap: cannot wrap empty data.')
+
+    def __repr__(self):
+        return 'Wrap(%s)' % ', '.join([('%s:%s' % (k, type(self[k]))) for k in sorted(self.keys())])
+
+    def __getattr__(self, key):
+        if key not in self:
+            raise AttributeError('Wrap: attribute "%s" does not exist.' % key)
+        return self[key]
+
+    def __hash__(self):
+        keys = sorted(self.keys())
+        types = []
+        attributes = []
+        for k in keys:
+            types += (type(self[k]),)
+            if isinstance(self[k], numpy.ndarray):
+                # Note: hash_from_ndarray returns a string, so the hash is not yet complete
+                # (__hash__ must return an integer).
+                attributes += (hash_from_ndarray(self[k]),)
+            else:
+                # No checking, data should be hashable.
+                attributes += (self[k],)
+        return hash((type(self),) + tuple(keys) + tuple(types) + tuple(attributes))
+
+    def __eq__(self, other):
+        if type(self) != type(other) or len(self) != len(other):
+            return False
+        for k in self:
+            if k not in other or not (isinstance(self[k], type(other[k])) and isinstance(other[k], type(self[k]))):
+                return False
+            if isinstance(self[k], numpy.ndarray):
+                if not numpy.allclose(self[k], other[k]):
+                    return False
+            elif self[k] != other[k]:
+                return False
```

Comment:
Ok ! I think it's better to keep both classes (to let Wrapper manage the types and Wrap manage the values), but I will modify Wrap so that the constructor takes as argument the Wrapper instance its depends on, and I'll use it in Wrap.__eq__ and Wrap.__hash__ as you suggest.